### PR TITLE
[Snowservices][User Metrics] Metrics Discovery Service

### PR DESCRIPTION
**Pull request by sfc-gh-tkommineni**
In this PR, We are providing a Metrics Endpoint Discovery Service.

This service provides list of endpoints that can be scraped for user metrics.

We have tested it by creating a spcs service

`CREATE SERVICE MDSERVICE
in compute pool MY_POOL
from specification '
spec:
  container:
  - name: otel
    image: docker.io/tkommineni/otel-mdservice:latest
  - name: mdservice
    image: docker.io/tkommineni/mdservice:latest
';
`

The result is as follows:

md service response when the otel container makes a call via http_sd_config:
Screenshot 2024-01-05 at 2 23 32 PM

otel collector scraping metrics successfully.
Screenshot 2024-01-05 at 2 24 09 PM

### CODE REVIEW

File: user-metrics/mdservice/config.yaml [hunk 1]
```diff
@@ -0,0 +1 @@
+port: 8080 # Port where the discovery service is hosted
```
Comment (sfc-gh-aivanou) [inline user-metrics/mdservice/config.yaml:1]: what about account, host, username, etc..
the same as it was in plugin?

File: user-metrics/mdservice/spcs.go [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+	RoleQuery           = "SELECT CURRENT_ROLE();"
+	NameColumn          = "name"
+	StateColumn         = "state"
+	OwnerColumn         = "owner"
+	StateActiveValue    = "active"
+	DnsPrefix           = "metrics"
+	DnsSuffix           = "snowflakecomputing.internal"
+	MetricsPort         = "9001"
+	MetaSpcsPrefix      = SpcsPluginName + "_"
+	MetaComputePoolName = "__meta_" + MetaSpcsPrefix + "compute_pool_name"
```
Comment (sfc-gh-aivanou) [inline user-metrics/mdservice/spcs.go:33]: simplify pls to a single const string. You don't use `MetaSpcsPrefix` and `SpcsPluginName` anywhere but here

File: user-metrics/mdservice/spcs.go [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+var config *Config
+
+// Config represents the structure of configuration file
+type Config struct {
+	Port int `yaml:"port"`
+}
+
+// EndpointResolver interface definition
+type EndpointResolver interface {
+	ResolveEndPoint(endPoint string) ([]string, error)
```
Comment (sfc-gh-aivanou) [inline user-metrics/mdservice/spcs.go:47]: nit: `ResolveEndpoint`

File: user-metrics/mdservice/spcs.go [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+}
+
+// DNSResolver is a real implementation of EndpointResolver
+type DNSResolver struct{}
+
```
Comment (sfc-gh-aivanou) [inline user-metrics/mdservice/spcs.go:52]: move methods that `DNSResolver` implements here

File: user-metrics/mdservice/spcs.go [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+func getSfConfig() (sf.Config, error) {
```
Comment (sfc-gh-aivanou) [inline user-metrics/mdservice/spcs.go:53]: pls move all snowflake-related functions to a separate file

File: user-metrics/mdservice/spcs.go [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+//   - cfg: An sf.Config struct containing Snowflake database configuration.
+//   - query: Query to run in Snowflake.
+//
+// Returns:
+//   - rows: A 2D string slice representing the result rows of the executed query.
+//   - columns: A string slice containing the names of the columns in the result set.
+//   - error: An error, if any, encountered during the database connection, query execution, or data retrieval.
+func getDataFromSnowflake(cfg sf.Config, query string) ([][]string, []string, error) {
+
+	// Create DSN from config
```
Comment (sfc-gh-aivanou) [inline user-metrics/mdservice/spcs.go:106]: you don't need these comments

File: user-metrics/mdservice/config.yaml (hunk 1 above)
Comment (sfc-gh-tkommineni) [reply user-metrics/mdservice/config.yaml:1]: the discovery service will only work as an spcs service and every spcs service has the oauth token associated with it. So I was wondering, why not keep it simple and use the token to execute queries.

I can add password based auth too

File: user-metrics/mdservice/spcs.go [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+	dsn, err := sf.DSN(&cfg)
+	if err != nil {
+		return nil, nil, fmt.Errorf("SPCS discovery plugin: failed to create DSN from Config. err: %v", err)
+	}
+
+	// Establish connection with snowflake
```
Comment (sfc-gh-aivanou) [inline user-metrics/mdservice/spcs.go:112]: also remove

File: user-metrics/mdservice/spcs.go [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+	db, err := sql.Open("snowflake", dsn)
+	if err != nil {
+		return nil, nil, fmt.Errorf("SPCS discovery plugin: failed to connect. err: %v", err)
+	}
+	defer db.Close()
+
+	// Execute query
```
Comment (sfc-gh-aivanou) [inline user-metrics/mdservice/spcs.go:119]: here as well

File: user-metrics/mdservice/spcs.go [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+	// Create DSN from config
+	dsn, err := sf.DSN(&cfg)
+	if err != nil {
+		return nil, nil, fmt.Errorf("SPCS discovery plugin: failed to create DSN from Config. err: %v", err)
+	}
+
+	// Establish connection with snowflake
+	db, err := sql.Open("snowflake", dsn)
+	if err != nil {
+		return nil, nil, fmt.Errorf("SPCS discovery plugin: failed to connect. err: %v", err)
```
Comment (sfc-gh-aivanou) [inline user-metrics/mdservice/spcs.go:115]: add host and account info to error msg

File: user-metrics/mdservice/spcs.go [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+	if err != nil {
+		return "", err
+	}
+	// Close the file.
+	defer file.Close()
+
+	return string(fileContents), nil
+}
+
+// getDataFromSnowflake establishes a connection with a Snowflake database, executes a query,
```
Comment (sfc-gh-aivanou) [inline user-metrics/mdservice/spcs.go:93]: why did you document only a single private function?

File: user-metrics/mdservice/spcs.go [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+		if err != nil {
+			logger.Printf("SPCS discovery plugin: Error getting compute pools to scrape. err: %v", err)
+		}
+		logger.Printf("SPCS discovery plugin: Compute Pools that can be scraped %s\n", computePools)
+
+		response, err := getResponse(computePools)
+		if err != nil {
+			logger.Printf("SPCS discovery plugin: Error getting targetgroups to scrape. err: %v", err)
+		}
+		logger.Printf("SPCS discovery plugin response: %s\n", response)
```
Comment (sfc-gh-aivanou) [inline user-metrics/mdservice/spcs.go:363]: pls remove print, and use debug

File: user-metrics/mdservice/spcs.go [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+func indexOf(target string, slice []string) int {
+	for i, value := range slice {
+		if value == target {
+			return i
+		}
+	}
+	return -1 // Element not found
+}
+
+func getResponse(computePools []string) ([]struct {
```
Comment (sfc-gh-aivanou) [inline user-metrics/mdservice/spcs.go:228]: rename this method to something that represents what this method does

File: user-metrics/mdservice/spcs.go [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+		targetForCP := fmt.Sprintf("%s:%s", ipAddress, MetricsPort)
+		targetsForCP = append(targetsForCP, targetForCP)
+	}
+
+	return targetsForCP, nil
+}
+
+func getLabels(computePool string) map[string]string {
+	labels := map[string]string{
+		MetaComputePoolName: strings.ToLower(computePool),
```
Comment (sfc-gh-aivanou) [inline user-metrics/mdservice/spcs.go:279]: you can add account as label as well

File: user-metrics/mdservice/config.yaml (hunk 1 above)
Comment (sfc-gh-aivanou) [reply user-metrics/mdservice/config.yaml:1]: okay, lets keep it as it is

**Review by sfc-gh-aivanou — APPROVED** [review]

**Pull request merged**